src/TestFramework/PhpUnit/Coverage/IndexXmlCoverageParser.php

```diff
@@ -77,29 +79,24 @@ private function parseNodes(string $coverageIndexPath, SafeDOMXPath $xPath): ite
     {
         $projectSource = self::getProjectSource($xPath);
 
-        $nodes = $xPath->query('//file');
-
-        foreach ($nodes as $node) {
+        foreach ($xPath->query('//file') as $node) {
             $relativeCoverageFilePath = $node->getAttribute('href');
 
-            $fileInfoProvider = new SourceFileInfoProvider(
-                $coverageIndexPath,
-                $this->coverageDir,
-                $relativeCoverageFilePath,
-                $projectSource
+            yield $this->xmlCoverageParser->parse(
+                new SourceFileInfoProvider(
+                    $coverageIndexPath,
+                    $this->coverageDir,
+                    $relativeCoverageFilePath,
+                    $projectSource
+                )
```

Comment:
There's no need for `IndexXmlCoverageParser` to be even aware of presence of `XmlCoverageParser`. `IndexXmlCoverageParser` produces `SourceFileInfoProvider`, and `XmlCoverageParser` consumes. They can be connected by some other entity, like `PhpUnitXmlCoveredFileDataProvider`.

Reply:
I'm not sure to understand you: `IndexXmlCoverageParser` is _already_ aware of `XmlCoverageParser`. The only change here is instead of making the later stateful, requiring it to be instantiated for each file info provider, it is now stateless

Reply:
It does, but it shouldn't be. It's was a legacy decision made just to make `IndexXmlCoverageParser` be compatible with old tests. `IndexXmlCoverageParser`'s only true job is to produce `SourceFileInfoProvider`s, hence we have them in the first place.

Reply:
I honestly fail to see the issue here: I'm not trying to get a step further and correcting a legacy or anything, but just switch a service from being stateful to stateless which avoids to instantiate it several times

Reply:
Well, then how could I guess that. We'll do the next later then.

Reply:
Yes sorry I should have tried to make it a bit more clear :)